crates/llvm-context/src/polkavm/evm/call.rs

```diff
@@ -103,20 +103,79 @@ where
 
 #[allow(clippy::too_many_arguments)]
 pub fn delegate_call<'ctx, D>(
-    _context: &mut Context<'ctx, D>,
+    context: &mut Context<'ctx, D>,
     _gas: inkwell::values::IntValue<'ctx>,
-    _address: inkwell::values::IntValue<'ctx>,
+    address: inkwell::values::IntValue<'ctx>,
     _value: Option<inkwell::values::IntValue<'ctx>>,
-    _input_offset: inkwell::values::IntValue<'ctx>,
-    _input_length: inkwell::values::IntValue<'ctx>,
-    _output_offset: inkwell::values::IntValue<'ctx>,
-    _output_length: inkwell::values::IntValue<'ctx>,
+    input_offset: inkwell::values::IntValue<'ctx>,
+    input_length: inkwell::values::IntValue<'ctx>,
+    output_offset: inkwell::values::IntValue<'ctx>,
+    output_length: inkwell::values::IntValue<'ctx>,
     _constants: Vec<Option<num::BigUint>>,
 ) -> anyhow::Result<inkwell::values::BasicValueEnum<'ctx>>
 where
     D: Dependency + Clone,
 {
-    todo!()
+    let address_pointer = context.build_address_argument_store(address)?;
+
+    let extcodehash_pointer =
+        context.build_alloca_at_entry(context.word_type(), "extcodehash_pointer");
+
+    context.build_runtime_call(
+        runtime_api::imports::CODE_HASH,
+        &[
+            address_pointer.to_int(context).into(),
+            extcodehash_pointer.to_int(context).into(),
+        ],
+    );
+
+    context.build_byte_swap(context.build_load(extcodehash_pointer, "extcodehash_value")?)?;
+
+    let input_offset = context.safe_truncate_int_to_xlen(input_offset)?;
+    let input_length = context.safe_truncate_int_to_xlen(input_length)?;
+    let output_offset = context.safe_truncate_int_to_xlen(output_offset)?;
+    let output_length = context.safe_truncate_int_to_xlen(output_length)?;
+
+    // TODO: What to supply here? Is there a weight to gas?
```

Comment:
We need some further investigation on this. Why is gas not a supported parameter in the pallet? Can it be supported? If so, that needs either an issue or PR for the pallet. Otherwise the gas argument should be removed from this function altogether and the reason documented.

Reply:
I think we should replace weight (ref_time, proof_size) with just gas on revive. Solidity has no concept of PoV and all weight (substrate) related operations should be handled by pallet-revive (gas to weight and so on)

Reply:
It cannot really be handled by pallet-revive alone. Since Solidity has no concept of it we have to assume some default limits for proof size and storage deposit limit. As of right now the default is just "use all of the remaining". This is the most compatible but doesn't allow proper limiting. The plan is to expose the extended interface via pre-compile to allow Solidity contracts to limit all dimensions.

That being said, we decided to leave all parameters on the host function and revive just fills them out with `0`. This way other languages can at least use them.

Reply:
Yes you are right, for Solidity, ideally we just pass through what we get and let the pallet handle it. For other languages however this might be quite beneficial since they can just decide to not ignore it. Mapping everything it into the 1D (ref_time) space implies wasting block-space. Which is especially undesired for other languages like Rust/ink! because we expect people to use those for implementing more compute heavy logic. So to me it seems that ideally we'd have both. Or, since this is a pointer anyways, it could point to something like `enum Weights { Gas(U256), Native { ref_time: u64, proof_size: u64 } }`

Reply:
@ermalkaleci FWIW, in this PR it's fine to set it to remove the argument and set it to 0. The whole EVM gas vs. substrate weights problem is it's own topic.

Reply:
Of course we can provide some additional API that just takes "gas". But this would just be optimization. It has exactly the same semantics as filling `0` as we do now.

Reply:
What if we replace `ref_time` with `gas` and let runtime do conversion to Weight (`proof_size` is optional, unused for solidity). Other contracts build without revive using Weight will have to convert `ref_time` to `gas`. At the end PolkaVM knows how to handle gas. `proof_size` remains optional, probably rename it to `storage_limit`

Reply:
I don't see how `ref_time` can be equal to `gas`. 1 instruction means 1 unit of gas but it doesn't mean it's 1 picosecond. For brand new chains not depending on benchmarked `ref_time` it's easy because you can can measure how much gas is consumed into 1 seconds and use that reference to limit gas per block, but in case of parachains they need to convert between `gas` and `ref_time`

Reply:
I propose we rename `ref_time` -> `gas_limit` and `proof_size` -> `storage_limit`

Reply:
> I don't see how ref_time can be equal to gas. 

It is just how we define gas. This is where we incompatible to EVM. We use different gas values to allow making certain operations cheaper than others.

Reply:
True but I don't see how we can match 1 unit of gas with 1 unit of time (we may come close though). Anyway renaming to `gas_limit` and `storage_limit` makes more sense to me.

Reply:
> True but I don't see how we can match 1 unit of gas with 1 unit of time (we may come close though)

We are doing this: ` 1 gas = 1 ref_time = 1 picosecond of execution on the reference machine`

Reply:
Why renaming it?

Reply:
Ok I see but `gas_limit` and `storage_limit` are terms that neither exist in EVM nor in Substrate :) I think that this would only increase the confusion.

Reply:
And where do you want to rename them? Here in the recompiler they obviously don't come up?

Reply:
Never mind. It's fine